corehq/util/tests/test_soft_assert.py

```diff
@@ -0,0 +1,128 @@
+        self.assertEqual(self.infos[2].key, key0, '\n{}\n{}'.format(
+            self.infos[0].traceback,
+            self.infos[2].traceback,
+        ))
+        self.assertEqual(self.infos[3].key, key1, '\n{}\n{}'.format(
+            self.infos[1].traceback,
+            self.infos[3].traceback,
+        ))
+        self.assertEqual(self.infos[0].line,
+                         'if not self.soft_assert(isinstance(x, float)):')
+        self.assertEqual(self.infos[1].line,
+                         'if not self.soft_assert(isinstance(x, float)):')
+        self.assertEqual(self.infos[2].line,
+                         'if not self.soft_assert(isinstance(x, float)):')
+        self.assertEqual(self.infos[3].line,
+                         'if not self.soft_assert(isinstance(x, float)):')
+
+
+class SoftAssertHelpersTest(SimpleTestCase):
+    def test_number_is_power_of_two(self):
+        powers_of_two = [2**i for i in range(10)]
+        for n in range(100):
+            actual = _number_is_power_of_two(n)
+            expected = n in powers_of_two
+            self.assertEqual(actual, expected,
+                             '_number_is_power_of_two: {}'.format(actual))
+
+    def test_send_message(self):
+
+        def test1(subject, message):
+            self.assertRegexpMatches(subject,
+                                     r"Soft Assert: \[\w+\] This should fail")
+            things_that_should_show_up_in_message = [
+                r"Message: This should fail",
+                r"Traceback:",
+                r'File ".*corehq/util/tests/test_soft_assert.py", line \d+, in test_send_message',
+                r"self\.soft_assert\(False, 'This should fail'\)",
+                r"Occurrences to date: 1",
+            ]
+            for thing in things_that_should_show_up_in_message:
+                self.assertRegexpMatches(
+                    message,
+                    thing,
+                    '{!r}\ndoes not match\n---\n{}---\n'.format(thing, message)
+                )
+
+        def test2(subject, message):
+            self.assertRegexpMatches(
+                subject,
+                r'Soft Assert: \[\w+\] None',
+            )
+            things_that_should_show_up_in_message = [
+                r"Message: None",
+                r"Traceback:",
+                r'File ".*corehq/util/tests/test_soft_assert.py", line \d+, in test_send_message',
+                r"self\.soft_assert\(False\)",
+                r"Occurrences to date: 1",
+            ]
+            for thing in things_that_should_show_up_in_message:
+                self.assertRegexpMatches(
+                    message,
+                    thing,
+                    '{!r}\ndoes not match\n---\n{}---\n'.format(thing, message)
+                )
+
+        tests = iter([test1, test2])
+
+        def backend(subject, message):
+            test_ = tests.next()
+            test_(subject, message)
+
+        self.soft_assert = SoftAssert(
```

Comment:
In the future django has a pretty good framework for testing emails. Could make this easier so you don't have to fake the backend. https://docs.djangoproject.com/en/1.7/topics/testing/tools/#email-services